File: src/libs/ThirdParty.VBSuppliers/Properties/Travelgate/Models/TravelgateSearchResponse.vb
Imports System.Xml.Serialization

<Serializable(), XmlRoot("Envelope")>
Public Class TravelgateResponseEnvelope
    <XmlElement("Body")> Public Property Body As TravelgateSearchEncodedResponse
End Class

<Serializable(), XmlType("Body")>
Public Class TravelgateSearchEncodedResponse
    <XmlElement("AvailResponse")> Public Property Response As ResponseDetails

    Public Class ResponseDetails
        <XmlElement("AvailResult")> Public Property Result As ProviderRSs
    End Class

    Public Class ProviderRSs
        <XmlElement("providerRSs")> Public Property ProviderResults As New List(Of ProviderResultsDetails)
    End Class

    Public Class ProviderResultsDetails
        <XmlElement("ProviderRS")> Public Property Results As ResultDetails
    End Class

    Public Class ResultDetails
        <XmlElement("rs")> Public Property Result As String
    End Class

End Class

<Serializable(), XmlRoot("AvailRS")>
Public Class TravelgateSearchResponse
    Public Property Hotels As New List(Of Hotel)
    <XmlArray("DailyRatePlans")>
    <XmlArrayItem("DailyRatePlan")>
    Public Property DailyRatePlans As New List(Of RatePlan)

    Public Class Hotel
        <XmlAttribute("code")> Public Property TPKey As String
        <XmlAttribute("name")> Public Property Name As String
        <XmlArray("MealPlans")>
        <XmlArrayItem("MealPlan")>
        Public Property Meals As New List(Of Meal)
    End Class

    Public Class Meal
        <XmlAttribute("code")> Public Property MealBaisCode As String
        <XmlArray("Options")>
        <XmlArrayItem("Option")>
        Public Property Options As New List(Of OptionDetails)
    End Class

    Public Class OptionDetails
        <XmlAttribute("type")> Public Property Type As String
        <XmlAttribute("paymentType")> Public Property PaymentType As String
        <XmlAttribute("status")> Public Property Status As String
        Public Property Rooms As New List(Of Room)
        <XmlElement("Price")> Public Property Price As PriceDetails
        Public Property Parameters As New List(Of Parameter)
        Public Property RateRules As New List(Of RateRule)
    End Class

    Public Class Room
        <XmlAttribute("id")> Public Property ID As String
        <XmlAttribute("code")> Public Property RoomTypeCode As String
        <XmlAttribute("description")> Public Property RoomType As String
        <XmlAttribute("roomCandidateRefId")> Public Property PropertyRoomBookingID As String
        <XmlAttribute("nonRefundable")> Public Property NonRefunfable As String = String.Empty
    End Class

    Public Class PriceDetails
        <XmlAttribute("currency")> Public Property Currency As String
        <XmlAttribute("amount")> Public Property Amount As String
        <XmlAttribute("binding")> Public Property Binding As String
        <XmlAttribute("commission")> Public Property Commission As String
    End Class

    Public Class Parameter
        <XmlAttribute("key")> Public Property Key As String
        <XmlAttribute("value")> Public Property Value As String
    End Class

    Public Class RateRule
        <XmlElement("Rules")> Public Property Rules As New List(Of Rule)
    End Class

    Public Class Rule
        <XmlAttribute("type")> Public Property RateType As String
    End Class

    Public Class RatePlan
        <XmlAttribute("code")> Public Property TPRateCode As String = String.Empty
    End Class
End Class

File: src/libs/iVectorOne.Suppliers/Travelgate/Models/TravelgateSearchResponse.cs
namespace iVectorOne.Suppliers.Travelgate.Models
{
    using System;
    using System.Collections.Generic;
    using System.Xml.Serialization;

    [Serializable()]
    [XmlRoot("Envelope")]
    public class TravelgateResponseEnvelope
    {
        [XmlElement("Body")]
        public TravelgateSearchEncodedResponse Body { get; set; }
    }

    [Serializable()]
    [XmlType("Body")]
    public class TravelgateSearchEncodedResponse
    {
        [XmlElement("AvailResponse")]
        public ResponseDetails Response { get; set; }

        public class ResponseDetails
        {
            [XmlElement("AvailResult")]
            public ProviderRSs Result { get; set; }
        }

        public class ProviderRSs
        {
            [XmlElement("providerRSs")]
            public List<ProviderResultsDetails> ProviderResults { get; set; } = new List<ProviderResultsDetails>();
        }

        public class ProviderResultsDetails
        {
            [XmlElement("ProviderRS")]
            public ResultDetails Results { get; set; }
        }

        public class ResultDetails
        {
            [XmlElement("rs")]
            public string Result { get; set; }
        }

    }

    [Serializable()]
    [XmlRoot("AvailRS")]
    public class TravelgateSearchResponse
    {
        public List<Hotel> Hotels { get; set; } = new List<Hotel>();
        [XmlArray("DailyRatePlans")]
        [XmlArrayItem("DailyRatePlan")]
        public List<RatePlan> DailyRatePlans { get; set; } = new List<RatePlan>();

        public class Hotel
        {
            [XmlAttribute("code")]
            public string TPKey { get; set; }
            [XmlAttribute("name")]
            public string Name { get; set; }
            [XmlArray("MealPlans")]
            [XmlArrayItem("MealPlan")]
            public List<Meal> Meals { get; set; } = new List<Meal>();
        }

        public class Meal
        {
            [XmlAttribute("code")]
            public string MealBaisCode { get; set; }
            [XmlArray("Options")]
            [XmlArrayItem("Option")]
            public List<OptionDetails> Options { get; set; } = new List<OptionDetails>();
        }

        public class OptionDetails
        {
            [XmlAttribute("type")]
            public string Type { get; set; }
            [XmlAttribute("paymentType")]
            public string PaymentType { get; set; }
            [XmlAttribute("status")]
            public string Status { get; set; }
            public List<Room> Rooms { get; set; } = new List<Room>();
            [XmlElement("Price")]
            public PriceDetails Price { get; set; }
            public List<Parameter> Parameters { get; set; } = new List<Parameter>();
            public List<RateRule> RateRules { get; set; } = new List<RateRule>();
        }

        public class Room
        {
            [XmlAttribute("id")]
            public string ID { get; set; }
            [XmlAttribute("code")]
            public string RoomTypeCode { get; set; }
            [XmlAttribute("description")]
            public string RoomType { get; set; }
            [XmlAttribute("roomCandidateRefId")]
            public string PropertyRoomBookingID { get; set; }
            [XmlAttribute("nonRefundable")]
            public string NonRefunfable { get; set; } = string.Empty;
        }

        public class PriceDetails
        {
            [XmlAttribute("currency")]
            public string Currency { get; set; }
            [XmlAttribute("amount")]
            public string Amount { get; set; }
            [XmlAttribute("binding")]
            public string Binding { get; set; }
            [XmlAttribute("commission")]
            public string Commission { get; set; }
        }

        public class Parameter
        {
            [XmlAttribute("key")]
            public string Key { get; set; }
            [XmlAttribute("value")]
            public string Value { get; set; }
        }

        public class RateRule
        {
            [XmlElement("Rules")]
            public List<Rule> Rules { get; set; } = new List<Rule>();
        }

        public class Rule
        {
            [XmlAttribute("type")]
            public string RateType { get; set; }
        }

        public class RatePlan
        {
            [XmlAttribute("code")]
            public string TPRateCode { get; set; } = string.Empty;
        }
    }
}